### DISCUSSION TITLE
Announcing AI Gateway

### DISCUSSION THREAD

**Discussion by subnetmarco** [Kong/kong#12569] [Announcements]
Hello everyone 👋 , today we announced a set of new **AI plugins** in the broader Kong Gateway 3.6 release. You can read the [full AI announcement here](https://konghq.com/blog/product-releases/announcing-kong-ai-gateway).

Essentially we can now use Kong Gateway to proxy AI traffic to cloud and self-hosted LLMs with a whole set of L7 AI Capabilities when it comes to prompts, observability and request/response transformation, and more. All other plugins can now be also used on top of upstream AI services to establish more advanced governance, security and traffic control patterns.

You can learn more about the new AI capabilities and even see demonstration videos at the [official product page](https://konghq.com/products/kong-ai-gateway).

I am opening this discussion thread to collect feedback and ideas around this new AI use-case, and looking for feedback and suggestions from the community.

**Update**: Kong AI Gateway 3.8 [has been announced](https://konghq.com/blog/product-releases/ai-gateway-3-8) with semantic caching, semantic guardrails, semantic routing and more.

**Comment by kandakji** [comment]
Hi Marco @subnetmarco , this is great. We'd be happy to contribute to this AI proxy to add support for Amazon Bedrock invoke model API in order to add it to the list of providers you support

**Reply by subnetmarco** [reply]
That would be awesome to have.

**Reply by tysoekong** [reply]
Hi @kandakji , I built the initial range of provider integrations to support this AI category.

Would really like to meet, so we can plan out adding support for the Bedrock Inference format! We want to enable as many providers and model variety as possible.

**Reply by subnetmarco** [reply]
Update: AWS Bedrock will be released in the upcoming v3.8 this summer, along with other LLM technologies.

**Comment by subnetmarco** [comment]
A new release of AI Gateway has been announced with streaming support, new plugins, richer analytics and more: https://konghq.com/blog/product-releases/ai-gateway-goes-ga

We are going to release AWS Bedrock and Gemini support in just a few weeks. 

The new analytics allows to extract all sort of metrics from your GenAI usage:

<img width="739" alt="Screenshot 2024-05-29 at 19 24 25" src="https://github.com/Kong/kong/assets/964813/275e4681-e0ba-44ee-b06b-ad019812d122">

**Comment by guleng** [comment]
@subnetmarco I have many LLM models for ai proxy, can I directly load balance and perform health checks on multiple models? I was unable to achieve load balancing for the LLM model in my actual Kong configuration

**Reply by subnetmarco** [reply]
We are going to be announcing new load balancing capabilities soon that will allow to:

* Load balance across multiple combinations of providers/models.
* Choose between `6 different` load balancing algorithms: 
  * Round robin
  * Weighted
  * Least busy (least number of ongoing calls)
  * Lowest usage (lowest TPM usage)
  * Lowest latency
  * Semantic Routing
* Implement failovers and retries.

Among many other features that we have been building for the past few months.

**Reply by guleng** [reply]
Do you have an approximate time?

**Reply by subnetmarco** [reply]
August/Sept.

**Reply by mepc36** [reply]
Hey @subnetmarco, is the August/September delivery date you mentioned here still accurate?

Thanks!

**Reply by subnetmarco** [reply]
Yes, second week of September.

**Reply by subnetmarco** [reply]
We [just announced Kong AI Gateway 3.8](https://konghq.com/blog/product-releases/ai-gateway-3-8) which includes more advanced load balancing algorithms, among many other semantic capabilities.

**Comment by abhijitherekar** [comment]
Hi @subnetmarco , I am from Acuvity AI , a genAI security startup and we would like to integrate with Kong AI gateway. 

Please, let me know what is the best way to reach out to you. 
Thanks

**Comment by Colstuwjx** [comment]
Hi team, 

May I ask if we have any MCP Gateway feature updates or roadmap for this ? Just like higress ai gateway did, it supports to proxy existing open apis and act as a MCP Server Gateway in front of upstream apis, see [blog post](https://www.alibabacloud.com/blog/the-utility-is-here-to-convert-existing-openapi-in-bulk-to-mcp-server_602147)  , any plan for Kong gateway to implement this kind of feature?

**Reply by subnetmarco** [reply]
Yes, we already support the ability to govern, secure and observe MCP traffic: https://konghq.com/blog/product-releases/securing-observing-governing-mcp-servers-with-ai-gateway

**Reply by Colstuwjx** [reply]
Hi @subnetmarco , thanks for your reply. However, from the blog post, I didn't see any features like configuring a http api in a MCP Server plugin like higress did, and also I've tried to search "MCP" keyword in plugin marketplace, still no record found: https://docs.konghq.com/hub/?search=mcp

May I know how to use this feature ? Or it's ONLY limited to a dedicated konnect MCP Server, which means we could ONLY use konnect MCP Server to govern kong gateway itself, rather than using kong gateway to proxy existing http apis (Kong as a rest-api-to-mcp-server proxy) or MCP Servers (Kong as a MCP Bridge)

Thanks.

**Reply by gperanich** [reply]
Hi @Colstuwjx, to expose a MCP server with Kong we would leverage the "rest-api-to-mcp-server proxy" as you mentioned above. 

In blog (+underlying application), we are using [native Kong Gateway routing](https://docs.konghq.com/gateway/latest/how-kong-works/routing-traffic/#routes-and-matching-capabilities) to expose the MCP server. This grants the ability to layer policy on the MCP server API. For example, in the blog we use the following plugins to secure and govern the MCP server exposed via Kong Gateway: 
- [OpenID Connect](https://docs.konghq.com/hub/kong-inc/openid-connect/) - to secure the MCP server with AuthN + AuthZ
- [HTTP Log](https://docs.konghq.com/hub/kong-inc/http-log/) - to integrate with our observability solution

Please have a look at the [Kong Plugin Hub](https://docs.konghq.com/hub/) to see the other types of policy that can be leveraged to manage MCP servers with Kong.

**Reply by Colstuwjx** [reply]
> we are using [native Kong Gateway routing](https://docs.konghq.com/gateway/latest/how-kong-works/routing-traffic/#routes-and-matching-capabilities) to expose the MCP server

But how could we expose those http services as MCP Server ? For the higress implementation, it supports to configure a http server endpoint in its MCP-Server wasm plugin (similar to Kong plugin), an example would be like this:

```
  matchRules:
  - config:
      server:
        name: random-user-server
      tools:
      - description: Get random user information
        name: get-user
        requestTemplate:
          method: GET
          url: https://fakerapi.it/api/v1/persons?_locale=en&_quantity=1
        responseTemplate:
          body: |-
            # User Information
            {{- with (index .data 0) }}
            - **Name**: {{.firstname}} {{.lastname}}
            - **Email**: {{.email}}
            - **Location**: {{.address.city}}, {{.address.country}}
            - **Phone**: {{.phone}}
            {{- end }}
    configDisable: false
    ingress:
    - user
```

By using this wasm plugin, higress could proxy the http api `https://fakerapi.it/api/v1/persons` and make it become a MCP Server. However, I didn't see any features like this in Kong, may I know how to achieve the same goal, for example, make a MCP Tool to invoke `https://fakerapi.it/api/v1/persons` api by using Kong as rest api to mcp server proxy ?

**Reply by subnetmarco** [reply]
To expose every API into an MCP server automatically will be an upcoming capability, it is not available yet.
